### DISCUSSION TITLE
New HWID system

### DISCUSSION THREAD

**Discussion by PJB3005** [space-wizards/space-station-14#33532] [Hosting Announcements]
The next stable release of the game will make use of a new HWID system. This is important to be aware of for server administration when pulling the latest version downstream.

If you want to manually cherry-pick this improvement, the relevant PRs are https://github.com/space-wizards/space-station-14/pull/33265 and https://github.com/space-wizards/space-station-14/pull/33461.

# New system

The new "modern" HWID system works by passing HWID values through the authentication server. This brings the following improvements:

* It **completely eliminates spoofing**, where one malicious player can intentionally pretend to have another's HWID to get them banned.
* It's more future proof, and will automatically work with all future improvements we are planning to make to the HWID system.
* It **DOES NOT** currently make HWIDs any more watertight against ban evasion than they were before.

**All new entries made (new bans, connection logs, etc.) will track the new HWID type.** In UI, these new HWIDs are displayed with a "`V2-`" prefix to visually distinguish them.

Old HWIDs are still available to game servers, so that the game server can match pre-existing bans. In the future, we may block ability to place new ban records for legacy HWIDs, to avoid accidents. Legacy HWIDs will not be improved and due to the risk of spoofing we consider them well and thoroughly broken.

# Database

The new system is accompanied by some database changes. The relevant migration to run, if you want to do it manually, is as follows:

```sql
START TRANSACTION;

ALTER TABLE server_role_ban ADD hwid_type integer DEFAULT 0;

ALTER TABLE server_ban ADD hwid_type integer DEFAULT 0;

ALTER TABLE player ADD last_seen_hwid_type integer DEFAULT 0;

ALTER TABLE connection_log ADD hwid_type integer DEFAULT 0;

INSERT INTO "__EFMigrationsHistory" ("MigrationId", "ProductVersion")
VALUES ('20241111170112_ModernHwid', '8.0.0');

COMMIT;

START TRANSACTION;

ALTER TABLE connection_log ADD trust real NOT NULL DEFAULT 0;

INSERT INTO "__EFMigrationsHistory" ("MigrationId", "ProductVersion")
VALUES ('20241111193608_ConnectionTrust', '8.0.0');

COMMIT;
```

The game servers will of course automatically apply this migration when necessary, so manual migration is only needed if you want to deploy the SS14.Admin changes in advance.

# SS14.Admin

SS14.Admin has been updated to version `1.6.0` to support the new HWID system. If you don't update, HWID values will display incorrectly/confusingly in the UI.

**Comment by jungarikjan** [comment]
I LOVE YOU

**Comment by ca4tuk** [comment]
It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
 It DOES NOT currently make HWIDs any more watertight against ban evasion than they were before.
